kvs: handle non-JSON values in flux kvs get, put, dir

garlick opened the pull request:
This PR updates the `flux kvs` command to handle non-JSON values.  These deficiencies were discussed  in #1158 and #1159.

A summary of the changes proposed here are:
* `flux kvs dir` displays non-JSON values
* `flux kvs dir` truncates output to fit terminal width, controlled with `--width=COLS`
* `flux kvs put` and `flux kvs get` now accept the `--raw`, `--json`, and `--treeobj` options to allow greater control over how values are stored and interpreted.

As a reminder the KVS just stores raw values now, imposing no structure.  We have convenience functions for get/put of strings (NULL terminated), and JSON (encoded as strings).  The get/put subcommands now default to storing strings, as opposed to JSON-encoded strings.  The `--json` option was  added to emulate the past behavior.  The `--raw` options was added to allow non-NULL terminated byte sequences to be stored and retrieved.  The `--treeobj` option was added so that snapshots could be manipulated and metadata could be examined conveniently.

Man pages updated and sharness tests added.

chu11 commented:
just beginning to review, but a thought that comes to mind, are many options in ```t/kvs/basic``` axe-able?  and tests adjusted accordingly?  copy-fromkvs, copy-tokvs, get-treeobj, put-treeobj, spring to mind.  Could be cleanup via another issue of course and not necessarily this PR.

garlick commented:
Hmm, it does seem like get-treeobj and put-treeobj are obviously the same as get/put with the --treeobj option.  I'll see if I can do some of that cleanup here.

chu11 commented on src/cmd/flux-kvs.c:
```diff
@@ -56,6 +56,16 @@ static void dump_kvs_dir (const flux_kvsdir_t *dir, bool Ropt, bool dopt);
 
```
typo in commit comment, "strong" -> "string"

chu11 replied:
hmm, comment appears to be attached to file and not a commit.  This is in commit https://github.com/flux-framework/flux-core/pull/1257/commits/f93f7768ffc5aef5901118bd48541c826d0c7966

garlick commented:
Just rebased on current master, fixed commit comment, and added some commits that move functionality out of `t/kvs/basic` that now seem appropriate in `flux-kvs`:
```
get [--at ref]        # replaces t/kvs/basic getat
dir [--at ref]        # replaces t/kvs/basic dirat
readlink [--at ref]   # replaces t/kvs/basic readlinkat
put --raw key=-       # (read from stdin), replaces t/kvs/basic copy-tokvs
put [--no-merge]      # replaces t/kvs/basic put-no-merge
```
Good suggestion @chu11.  All that's left in the test command is the type checking stuff.

chu11 commented on t/t1000-kvs.t:
```diff
@@ -438,6 +438,37 @@ test_expect_success 'kvs: zero-length value does not cause ls -FR to fail' '
+	flux kvs put $DIR.valref=$(seq -s 1 100) &&
+	flux kvs mkdir $DIR.dirref &&
+	flux kvs link foo $DIR.symlink &&
+	flux kvs get --treeobj $DIR.val &&
```
perhaps ```| grep "dir"``` and equivalents should be added to make sure the right treeobj is being returned

garlick replied:
Sure, that's a good improvement.

chu11 commented on t/t1000-kvs.t:
```diff
@@ -438,6 +438,39 @@ test_expect_success 'kvs: zero-length value does not cause ls -FR to fail' '
+test_expect_success 'kvs: empty string value does not cause dir to fail' '
+	flux kvs unlink -Rf $DIR &&
+	flux kvs put $DIR.a= &&
+	flux kvs dir $DIR
```
should this test (and the ls one below) also make sure the directory is output?

garlick replied:
Yep, will fix that.

chu11 commented:
@garlick awesome.  A bunch of the tests can probably be moved from ```t1002-kvs-extra.t``` to the main ```t1000-kvs.t``` now that they are more "core" tests.

chu11 commented:
just finished skimming.  Nothing stands out to me as a problem.

Before your rebase, it felt like there were a lack of basic ```--raw``` tests.  There are some now due to ```copy-fromkvs``` and ```copy-tokvs``` being gone, but some basic ```--raw``` functionality tests would probably be good.

A good test I thought of was to mix treeobj & raw.  Off the top of my head:

```
flux kvs put --raw $DIR.a={ a tree object val }
flux kvs get --raw $DIR.a (get the treeobj above)
flux kvs get --treeobj $DIR.a (can get a treeobj)
flux kvs get $DIR.a (can get value in the treeobj)
```

chu11 commented on t/t1000-kvs.t:
```diff
@@ -396,26 +396,132 @@ test_expect_success 'kvs: ls key. fails if key does not exist' '
+	flux kvs put $DIR.a.b=goodbye &&
+	flux kvs put --treeobj $DIR.b=$(flux kvs get --treeobj $DIR.a) &&
+	(flux kvs get $DIR.a.a && flux kvs get $DIR.a.b) >snap.expected &&
+	(flux kvs get $DIR.b.a && flux kvs get $DIR.b.b) >snap.actual
```
chain-lint needs &&?

garlick replied:
Good catch!

garlick commented:
Just pushed some changes to address your comments, except I didn't add any new raw tests or move any tests between the two sharness scripts.

I agree the test scripts are a little haphazard in the way they've come accreted tests over time and could use a cleanup pass (possibly cleanup is for another time though?) 

I'll go through the raw tests including the converted ones tomorrow and see if there are any obvious gaps.

garlick commented:
restarted one builder after hang in the cron tests

chu11 commented:
agreed, we should cleanup the tests another day.  After this PR is done, we can make an issue for it.  I'm sure after append support, the tests haphazardness will just be weirder.  We may want to split it up amongst even more files instead of just "core" and "extra".

garlick commented:
I didn't entirely understand what was meant here - maybe I am misreading
```
flux kvs put --raw $DIR.a={ a tree object val }
flux kvs get --raw $DIR.a (get the treeobj above)
flux kvs get --treeobj $DIR.a (can get a treeobj)
flux kvs get $DIR.a (can get value in the treeobj)
```
the first command would encode { tree object val } _inside_ a tree object val, the second would pull it out again.  Is nesting those meaningful?

On reorganizing - a good start would be to separate the kvs_watch tests out to their own script since they are a bit complicated and are likely to change when watch is reworked.

chu11 commented:
@garlick ahh, no nesting intended.  I guess I shouldn't have used brackets.  First line is to just write a treeobj, second line to read it back and make sure it wrote correctly.  I also realize now that the 4th line probably should have been with ```--json```.  Ugh I really messed up that suggestion.

garlick commented:
I've pushed some raw tests, mainly exercising the `kvs put --raw` command line. 

To be clear, --treeobj gives direct access to the RFC 11 object and --raw  creates a 'val' object just like --json  or no options.   I'm still confused by the suggestion - maybe we should chat face to face.

garlick commented:
Squashed.

chu11 commented on src/cmd/flux-kvs.c:
```diff
@@ -304,37 +309,85 @@ int main (int argc, char *argv[])
+    char *val, *kv, *p;
+    bool overflow = false;
+
+    assert (maxcol == 0 || maxcol > 3);
```
just realized, I don't think there's a check for a reasonable maxcol in callers of this function.  Should a check be added somewhere?  Or perhaps there should just be a generic "can't output with such small width" kind of error message?

garlick replied:
Yes you're right.

chu11 commented on doc/man1/flux-kvs.adoc:
```diff
@@ -42,13 +42,21 @@ arguments are described below.
+to be relative to an RFC 11 snapshot reference.
 
-*put* 'key=value' ['key=value...']::
+*put* [-j|-r] 'key=value' ['key=value...']::
```
need -t and -n

garlick replied:
I could swear I already did that one.  Thanks!

chu11 commented on doc/man1/flux-kvs.adoc:
```diff
@@ -42,13 +42,21 @@ arguments are described below.
+If '-j', it is first encoded as JSON, then stored as a NULL-terminated string.
+If '-r', it is stored as raw data with no termination.  For raw mode only,
+a value of "-" indicates that the value should be read from standard input.
+If '-t', value is stored as a RFC 11 object.
```
and description of -n.

garlick replied:
Right!  Adding that.

chu11 commented:
other than those nits above, I think things look good.  Did you want to squash?

garlick commented:
Thanks!  Re-squashed.

chu11 commented:
everything, lgtm!
